### PR TITLE
Iteration using Richardson number (definition via Bulk-Ri and SCFs) 

### LINKED ISSUE #111: Modify iteration method to use Richardson-number+zeta stability parameter definitions
$\frac{\partial Ri_{b}}{\partial \zeta} = \frac{Ri_{b}}{\zeta}\Big( 1+ \Big[ \phi_{c}(\zeta) - \phi_c(\frac{\zeta z_{0c}}{z-d})\Big] F_{c}^{-1}(\zeta) - 2 \Big[ \phi_{m}(\zeta) - \phi_{m}(\frac{\zeta z_{0m}}{z-d})\Big] F_{m}(\zeta)^{-1} \Big)$,

where

$F_{c} = \Big[\log(\frac{z-d}{z_{0c}}) - \psi_{c}(\zeta) + \psi_{c}(\frac{z_{0c}\zeta}{z-d}) \Big]$
$F_{m} = \Big[\log(\frac{z-d}{z_{0m}}) - \psi_{m}(\zeta) + \psi_{m}(\frac{z_{0m}\zeta}{z-d}) \Big]$
$d$ is the displacement height of canopies, 

and 

$Ri_{b} = \frac{gz \Delta \mathrm{DSE_{v}}}{\mathrm{DSE_{v,in} |u|^2}}$,
$Ri_{b} = \zeta F_{c} F_{m}^{-2}$,
$\Delta(\cdot)$ is the difference between surface and interior property values.

The last two expressions can be used to iterative over the ($Ri_{b}, \zeta$) pair, with a known derivative expression $\frac{\partial Ri_{b}}{\partial \zeta}$ as shown above. $\phi_{c}$, $\phi_{m}$ are similarity functions for scalar($c$), momentum($m$) variables respectively, with definitions from the `UniversalFunctions.jl` module. 

$\psi_{c,m}$ are stability correction functions,

$\psi(\zeta) = \int_{0}^{\zeta} \frac{1-\phi(\zeta\prime)}{\zeta\prime} d\zeta\prime$

### PR DESCRIPTION
**Pull request by akshaysridhar**
Closes #111

### CODE REVIEW

File: src/SurfaceFluxes.jl [hunk 1]
```diff
@@ -288,7 +288,7 @@ function surface_conditions(
     param_set::APS,
     sc::AbstractSurfaceConditions{FT},
     scheme::SolverScheme = FVScheme();
-    tol::RS.AbstractTolerance = RS.RelativeOrAbsoluteSolutionTolerance(FT(3e-5), FT(Δz(sc) / 50)),
+    tol::RS.AbstractTolerance = RS.RelativeOrAbsoluteSolutionTolerance(100eps(FT), 10eps(FT)),
```
Comment (tapios) [inline src/SurfaceFluxes.jl:291]: These should not be FP dependent. A tolerance of something like 1e-2 on the Richardson number should suffice, independent of FP.

**Review by tapios — APPROVED** [review]

File: src/SurfaceFluxes.jl [hunk 2]
```diff
@@ -351,6 +351,51 @@ obukhov_length(sfc::SurfaceFluxConditions) = sfc.L_MO
...
+function compute_∂Ri∂ζ(param_set, sc::AbstractSurfaceConditions{FT}, uft, scheme, ζ::FT) where {FT}
+    # In this design, this ∂Ri∂ζ function is intended to be an 
+    # internal function to support the Newton iteration scheme
+    thermo_params = SFP.thermodynamics_params(param_set)
+    ufparams = SFP.uf_params(param_set)
+    grav = SFP.grav(param_set)
+    cp_d = SFP.cp_d(param_set)
+    DSEᵥ_in = TD.virtual_dry_static_energy(thermo_params, ts_in(sc), grav * z_in(sc))
+    DSEᵥ_sfc = TD.virtual_dry_static_energy(thermo_params, ts_sfc(sc), grav * z_sfc(sc))
+    𝓏 = Δz(sc)
```
Comment (szy21) [inline src/SurfaceFluxes.jl:368]: It's probably better to use something other than 𝓏. (I don't even know how to type it:)

File: src/SurfaceFluxes.jl [hunk 3]
```diff
@@ -401,13 +446,13 @@ function obukhov_length(
...
+        function root_ζ(ζ)
+            residual = compute_richardson_number(sc, DSEᵥ_in, DSEᵥ_sfc, grav) - compute_Ri_b(param_set,sc, uft, scheme, non_zero(ζ))
             return residual
         end
-        L_MO_init = FT(-1)
-        sol = RS.find_zero(root_l_mo, RS.NewtonsMethodAD(L_MO_init), soltype, tol, maxiter)
-        L_MO = sol.root
+        ζ₀ = FT(sign(ΔDSEᵥ))
+        sol = RS.find_zero(root_ζ, RS.NewtonsMethodAD(ζ₀), soltype, tol, maxiter)
+        L_MO = sol.root * Δz(sc)
```
Comment (szy21) [inline src/SurfaceFluxes.jl:455]:
```suggestion
        L_MO = Δz(sc) / sol.root
```
It should be like this I think?

Comment (akshaysridhar) [reply src/SurfaceFluxes.jl]: yup zeta = z/Lmo. typo fixed.

File: src/SurfaceFluxes.jl (hunk 2 above)
Comment (akshaysridhar) [reply src/SurfaceFluxes.jl]: \scrz -> dz

File: src/SurfaceFluxes.jl [hunk 4]
```diff
@@ -353,12 +354,64 @@ obukhov_length(sfc::SurfaceFluxConditions) = sfc.L_MO
 
 non_zero(v::FT) where {FT} = abs(v) < eps(FT) ? v + sqrt(eps(FT)) : v
 
+
+function compute_richardson_number(sc::AbstractSurfaceConditions{FT}, DSEᵥ_in, DSEᵥ_sfc, grav) where {FT}
+    return (grav * z_in(sc) * (DSEᵥ_in - DSEᵥ_sfc)) / (DSEᵥ_in * (windspeed(sc))^2)
```
Comment (szy21) [inline src/SurfaceFluxes.jl:359]: Should it be `z_in` or `Δz` in this equation?

File: src/SurfaceFluxes.jl [hunk 5]
```diff
@@ -353,12 +354,64 @@ obukhov_length(sfc::SurfaceFluxConditions) = sfc.L_MO
...
+    F_h = log(dz / z0h) - ψₕ + ψₕ₀
+    ∂Ri∂ζ = compute_Ri_b(param_set, sc, uft, scheme, ζ)/ζ*(1+1/F_h*(ϕₕ-ϕₕ₀)-2*(ϕₘ-ϕₘ₀)*(1/F_m))
+end
+
+function compute_Ri_b(param_set, sc::AbstractSurfaceConditions{FT}, uft, scheme, ζ) where {FT}
+    #TODO: Fix code duplication
+    # In this design, if a user knows ζ, they may use it 
+    # to compute the bulk Richardson number. This may be used 
+    # to compute the Lₘₒ based on the definition of the Bulk-Richardson number 
+    # in terms of the virtual dry static energy
```
Comment (szy21) [inline src/SurfaceFluxes.jl:389]: Should we fix the TODO here?

File: src/SurfaceFluxes.jl [hunk 6]
```diff
@@ -353,12 +356,64 @@ obukhov_length(sfc::SurfaceFluxConditions) = sfc.L_MO
 
 non_zero(v::FT) where {FT} = abs(v) < eps(FT) ? v + sqrt(eps(FT)) : v
 
+
+function compute_richardson_number(sc::AbstractSurfaceConditions{FT}, DSEᵥ_in, DSEᵥ_sfc, grav) where {FT}
```
Comment (charleskawczynski) [inline src/SurfaceFluxes.jl:360]:
```suggestion
function compute_richardson_number(sc::AbstractSurfaceConditions, DSEᵥ_in, DSEᵥ_sfc, grav)
```
we don't need to specialize on `FT` here if it's not being used

File: src/SurfaceFluxes.jl (hunk 5 above)
Comment (akshaysridhar) [reply src/SurfaceFluxes.jl]: Resolved in latest commit

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
bors try

**Comment by akshaysridhar** [comment]
bors try

**Pull request merged**